### PR TITLE
[CALCITE-4333] The Sort rel should be decorrelated even though it has…

### PR DESCRIPTION
**Pull request by danny0405**
… fetch or limit when its parent is not a Correlate

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java [hunk 1]
```diff
@@ -680,6 +677,17 @@ private static void shiftMapping(Map<Integer, Integer> mapping, int startIndex,
 
   public Frame getInvoke(RelNode r, RelNode parent) {
     final Frame frame = dispatcher.invoke(r);
+    if (frame != null && parent instanceof Correlate && r instanceof Sort) {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:680]: Does this code check only the parent or any ancestor? Is it valid to check only the parent?

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:680]: My understanding is that only when the sort is an input of Correlate, we can not decorrelate the sort's input (when the sort input has be decorrelated).

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:680]: Hmm, we should check all its ancestor, but i have no good way for this ? Should i put a flag there (indicating whether we are under a Correlate now ?) ?

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java]: Finally i changes to check if the sort has a parent, if it does, that means the sort is definitely a branch of `Correlate` (if the sort input has been rewritten)

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java]: Cool, thanks for checking!

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java]: I think this discussion is relevant for https://issues.apache.org/jira/browse/CALCITE-4437

**Pull request merged**